File: plugins/plugin-discord/typescript/providers/channelState.ts
import type { IAgentRuntime, Memory, Provider, State } from "@elizaos/core";
import { ChannelType } from "@elizaos/core";
import type { GuildChannel } from "discord.js";
import { requireProviderSpec } from "../generated/specs/spec-helpers";
import type { DiscordService } from "../service";
import { ServiceType } from "../types";

const spec = requireProviderSpec("channelState");

/**
 * Represents a provider for retrieving channel state information.
 * @type {Provider}
 * @property {string} name - The name of the channel state provider.
 * @property {Function} get - Asynchronous function that retrieves channel state information based on the provided runtime, message, and optional state parameters.
 * @param {IAgentRuntime} runtime - The agent runtime.
 * @param {Memory} message - The message object.
 * @param {State} [state] - Optional state object.
 * @returns {Promise<Object>} A promise that resolves to an object containing channel state data, values, and text.
 */
export const channelStateProvider: Provider = {
  name: spec.name,
  get: async (runtime: IAgentRuntime, message: Memory, state: State) => {
    const room = state.data?.room ?? (await runtime.getRoom(message.roomId));
    if (!room) {
      throw new Error("No room found");
    }

    // if message source is not discord, return
    if (message.content.source !== "discord") {
      return {
        data: {},
        values: {},
        text: "",
      };
    }

    const agentName = state?.agentName || "The agent";
    const senderName = state?.senderName || "someone";

    let responseText = "";
    let channelType = "";
    let serverName = "";
    const channelId = room.channelId ?? "";

    if (room.type === ChannelType.DM) {
      channelType = "DM";
      responseText = `${agentName} is currently in a direct message conversation with ${senderName}. ${agentName} should engage in conversation, should respond to messages that are addressed to them and only ignore messages that seem to not require a response.`;
    } else {
      channelType = "GROUP";

      if (!channelId) {
        runtime.logger.error(
          {
            src: "plugin:discord:provider:channelState",
            agentId: runtime.agentId,
            roomId: room.id,
          },
          "No channel ID found"
        );
        return {
          data: {
            room,
            channelType,
          },
          values: {
            channelType,
          },
          text: "",
        };
      }

      const discordService = runtime.getService(ServiceType.DISCORD) as DiscordService;
      if (!discordService) {
        runtime.logger.warn(
          {
            src: "plugin:discord:provider:channelState",
            agentId: runtime.agentId,
            channelId,
          },
          "No discord client found"
        );
        return {
          data: {
            room,
            channelType,
            channelId,
          },
          values: {
            channelType,
            channelId,
          },
          text: "",
        };
      }

      // Look up guild via channel instead of serverId (which is now a UUID)
      // Try cache first, then fetch if not cached (handles cold start / partial cache scenarios)
      let channel = discordService.client?.channels.cache.get(channelId) as
        | GuildChannel
        | undefined;
      if (!channel && discordService.client) {
        try {
          channel = (await discordService.client.channels.fetch(channelId)) as
            | GuildChannel
            | undefined;
        } catch (fetchError) {
          runtime.logger.debug(
            {
              src: "plugin:discord:provider:channelState",
              agentId: runtime.agentId,
              channelId,
              error: fetchError instanceof Error ? fetchError.message : String(fetchError),
            },
            "Failed to fetch channel"
          );
        }
      }
      const guild = channel?.guild;
      if (!guild) {
        runtime.logger.warn(
          {
            src: "plugin:discord:provider:channelState",
            agentId: runtime.agentId,
            channelId,
          },
          "Guild not found for channel (not in cache and fetch failed)"
        );
        return {
          data: {
            room,
            channelType,
            channelId,
          },
          values: {
            channelType,
            channelId,
          },
          text: "",
        };
      }
      serverName = guild.name;

      responseText = `${agentName} is currently having a conversation in the channel \`#${channel?.name || channelId}\` in the server \`${serverName}\``;
      responseText += `\n${agentName} is in a room with other users and should be self-conscious and only participate when directly addressed or when the conversation is relevant to them.`;
    }

    return {
      data: {
        room,
        channelType,
        serverName,
        channelId,
      },
      values: {
        channelType,
        serverName,
        channelId,
      },
      text: responseText,
    };
  },
};

export default channelStateProvider;

File: plugins/plugin-slack/python/elizaos_plugin_slack/providers/channel_state.py
"""
Channel state provider for Slack.
"""

from typing import Any, Dict, Optional

from ..types import SLACK_SERVICE_NAME, get_slack_channel_type


async def get_channel_state(runtime: Any, message: Any, state: Any) -> Dict[str, Any]:
    """Get the current Slack channel state."""
    room = state.data.get("room") if state and state.data else None
    if not room:
        room = await runtime.get_room(message.room_id)
    
    if not room:
        return {"data": {}, "values": {}, "text": ""}
    
    # If message source is not slack, return empty
    if message.content.get("source") != "slack":
        return {"data": {}, "values": {}, "text": ""}
    
    agent_name = state.agent_name if state else "The agent"
    sender_name = state.sender_name if state else "someone"
    
    response_text = ""
    channel_type = ""
    workspace_name = ""
    channel_name = ""
    channel_id = room.channel_id or ""
    thread_ts = room.metadata.get("thread_ts") if room.metadata else None
    
    slack_service = runtime.get_service(SLACK_SERVICE_NAME)
    if not slack_service or not slack_service.client:
        return {
            "data": {"room": room, "channel_type": "unknown", "channel_id": channel_id},
            "values": {"channel_type": "unknown", "channel_id": channel_id},
            "text": "",
        }
    
    # Get channel info
    channel = await slack_service.get_channel(channel_id) if channel_id else None
    
    if channel:
        channel_name = channel.name
        slack_channel_type = get_slack_channel_type(channel)
        
        if slack_channel_type == "im":
            channel_type = "DM"
            response_text = f"{agent_name} is currently in a direct message conversation with {sender_name} on Slack. {agent_name} should engage in conversation, responding to messages that are addressed to them."
        elif slack_channel_type == "mpim":
            channel_type = "GROUP_DM"
            response_text = f"{agent_name} is currently in a group direct message on Slack. {agent_name} should be aware that multiple people can see this conversation."
        else:
            channel_type = "PRIVATE_CHANNEL" if slack_channel_type == "group" else "PUBLIC_CHANNEL"
            
            if thread_ts:
                response_text = f"{agent_name} is currently in a thread within the channel #{channel_name} on Slack."
                response_text += f"\n{agent_name} should keep responses focused on the thread topic and be mindful of thread etiquette."
            else:
                response_text = f"{agent_name} is currently having a conversation in the Slack channel #{channel_name}."
                response_text += f"\n{agent_name} is in a channel with other users and should only participate when directly addressed or when the conversation is relevant to them."
            
            if channel.topic and channel.topic.value:
                response_text += f"\nChannel topic: {channel.topic.value}"
            if channel.purpose and channel.purpose.value:
                response_text += f"\nChannel purpose: {channel.purpose.value}"
    else:
        channel_type = "unknown"
        response_text = f"{agent_name} is in a Slack conversation but couldn't retrieve channel details."
    
    # Add workspace context if available
    team_id = slack_service.get_team_id()
    if team_id and room.world_id:
        world = await runtime.get_world(room.world_id)
        if world:
            workspace_name = world.name
            response_text += f"\nWorkspace: {workspace_name}"
    
    # Add thread context if applicable
    if thread_ts:
        response_text += f"\nThis is a threaded conversation (thread timestamp: {thread_ts})."
    
    return {
        "data": {
            "room": room,
            "channel_type": channel_type,
            "workspace_name": workspace_name,
            "channel_name": channel_name,
            "channel_id": channel_id,
            "thread_ts": thread_ts,
            "is_thread": bool(thread_ts),
            "topic": channel.topic.value if channel and channel.topic else None,
            "purpose": channel.purpose.value if channel and channel.purpose else None,
            "is_private": channel.is_private if channel else None,
            "num_members": channel.num_members if channel else None,
        },
        "values": {
            "channel_type": channel_type,
            "workspace_name": workspace_name,
            "channel_name": channel_name,
            "channel_id": channel_id,
            "is_thread": bool(thread_ts),
        },
        "text": response_text,
    }


channel_state_provider = {
    "name": "slackChannelState",
    "description": "Provides information about the current Slack channel context",
    "get": get_channel_state,
}
